core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushJoinIntoTableScan.java

```diff
@@ -0,0 +1,270 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import io.trino.matching.Capture;
+import io.trino.matching.Captures;
+import io.trino.matching.Pattern;
+import io.trino.metadata.Metadata;
+import io.trino.metadata.TableHandle;
+import io.trino.spi.connector.ColumnHandle;
+import io.trino.spi.connector.JoinApplicationResult;
+import io.trino.spi.connector.JoinCondition;
+import io.trino.spi.connector.JoinType;
+import io.trino.spi.expression.Variable;
+import io.trino.spi.predicate.TupleDomain;
+import io.trino.sql.ExpressionUtils;
+import io.trino.sql.analyzer.FeaturesConfig.JoinPushdownMode;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.planner.iterative.Rule;
+import io.trino.sql.planner.plan.JoinNode;
+import io.trino.sql.planner.plan.Patterns;
+import io.trino.sql.planner.plan.TableScanNode;
+import io.trino.sql.tree.BooleanLiteral;
+import io.trino.sql.tree.ComparisonExpression;
+import io.trino.sql.tree.Expression;
+import io.trino.sql.tree.SymbolReference;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static io.trino.SystemSessionProperties.getJoinPushdownMode;
+import static io.trino.matching.Capture.newCapture;
+import static io.trino.sql.ExpressionUtils.and;
+import static io.trino.sql.ExpressionUtils.extractConjuncts;
+import static io.trino.sql.planner.plan.Patterns.tableScan;
+import static java.util.Objects.requireNonNull;
+
+public class PushJoinIntoTableScan
+        implements Rule<JoinNode>
+{
+    private static final Capture<TableScanNode> LEFT_TABLE_SCAN = newCapture();
+    private static final Capture<TableScanNode> RIGHT_TABLE_SCAN = newCapture();
+
+    private static final Pattern<JoinNode> PATTERN =
+            Patterns.join()
+                    .with(Patterns.Join.left().matching(tableScan().capturedAs(LEFT_TABLE_SCAN)))
+                    .with(Patterns.Join.right().matching(tableScan().capturedAs(RIGHT_TABLE_SCAN)));
+
+    private final Metadata metadata;
+
+    public PushJoinIntoTableScan(Metadata metadata)
+    {
+        this.metadata = metadata;
+    }
+
+    @Override
+    public Pattern<JoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(JoinNode joinNode, Captures captures, Context context)
```

Comment:
Ensure this doesn't fire when one of the TS-es is 'forDelete'.
(i'd write `verify` becase there cannot be a forDelete + Join, but Martin would rather return empty :)